### PR TITLE
Fail _search request with trailing tokens

### LINKED ISSUE #28995: Rest _search endpoint accepts invalid json body
This issue was first reported in https://github.com/elastic/elasticsearch/issues/28985.
Any text after the last closing bracket is simply ignored by the request json parser. So for instance a request like:
```
{
  "query": {
          "term": {
            "ProductID": "one"
          }
  }
},
"field_after_last_bracket": {
}
}
```
... is accepted.
I don't know why we have this leniency which is why I am opening this issue as a bug.

### PR DESCRIPTION
**Pull request by jimczi**
This change validates that the `_search` request does not have trailing
tokens after the main object and fails the request with a parsing exception otherwise.

Closes #28995

### CODE REVIEW

**Review by jpountz — APPROVED** [review]
Looks good. Let's turn the error into a deprecation log on 6.x?

File: server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java [hunk 1]
```diff
@@ -951,12 +951,18 @@ private SearchSourceBuilder shallowCopy(QueryBuilder queryBuilder, QueryBuilder
         return rewrittenBuilder;
     }
 
+    public void parseXContent(XContentParser parser) throws IOException {
+        parseXContent(parser, false);
```
Comment (colings86) [inline server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java]: Should this default not be the other way around (pass in true by default)? The only time I think we should allow trailing tokens is in the _msearch API which is a special case. It seems safer to me to check and error on trailing tokens in the default case and then have the _msearch API use the more specialised method below?

**Review by colings86 — APPROVED** [review]
LGTM

**Comment by jimczi** [comment]
Thanks @jpountz @colings86 
I'll turn the error into a deprecation log for the backport in 6x.

**Pull request merged**